### PR TITLE
Typo suggestion for a variable with a name similar to struct fields

### LINKED ISSUE #97133: Improve the error message of typos on variables that has the same name as struct fields
If a method's arguments have a name close to a struct field and we use the field name by mistake when trying to refer to the variable, the message is not very helpful. This should be more common inside `new` methods.

Given the following code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fb2f6006c85d85a96e3440f9713c91cf)

```rust
struct Foo {
    config: String,
}

impl Foo {
    fn new(cofig: String) -> Self {
        // ^^^^^ typo
        Self { config }
    }

    fn do_something(cofig: String) {
        println!("{config}");
        //         ^^^^^^ typo (?)
    }
}
```

The current output is:

```
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find value `config` in this scope
 --> src/lib.rs:8:16
  |
8 |         Self { config }
  |                ^^^^^^ a field by this name exists in `Self`

error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find value `config` in this scope
  --> src/lib.rs:12:20
   |
12 |         println!("{config}");
   |                    ^^^^^^ a field by this name exists in `Self`
```

Compare the `new` error message to a free-function-style `new`:

Ideally, the output of the `new` should look like this:

```console
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find value `config` in this scope
  --> src/lib.rs:19:11
   |
19 |     Foo { config }
   |           ^^^^^^ help: a local variable with a similar name exists: `cofig`
```

Is quite not there yet, but is much more helpful as it points to a typo I didn't know I have. In this case, we could either the message pointing to me that I should write `Foo { config: cofig }`, or, ideally, pointing that `cofig` may be a typo to `config`.

The second case (`do_something`), is more complicated as It's not clear whether `cofig` is related to `config` or not.

Sorry if this is duplication, I couldn't find a similar issue 😅 .

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #97133

### CODE REVIEW

**Comment by compiler-errors** [comment]
Er, can I take some time to review this @Dylan-DPC?

@bors r-

**Comment by compiler-errors** [comment]
Anyways, this looks mostly good, except is there a reason you only suggest if we have a  matching field?

This is a case where we have a matching method, but local variable also matches:

```rust
struct Foo;

trait Config {
    fn config(self);
    fn something(self);
}

impl Config for Foo {
    fn config() {}

    fn something(self) {
        let cofig = 1;
        println!("{cofig}");
    }
}
```

What does the blessed tests look like with the suggestion moved to like line 500, where the old code would `return`?

File: src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+error[E0425]: cannot find value `config` in this scope
+  --> $DIR/typo-suggestion-for-variable-with-name-similar-to-struct-field.rs:15:20
+   |
+LL |         println!("{config}");
+   |                    ^^^^^^
+   |
+help: you might have meant to use the available field
+   |
+LL |         println!("{self.config}");
```
Comment (Kobzol) [inline src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr:27]: Drive-by comment: this probably shouldn't be emitted, as it won't compile.

Comment (compiler-errors) [reply src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr:27]: Yeah, I think that's a bit orthogonal to this diff, though. It might be possible to detect this case from HIR though. 

@TaKO8Ki, I would appreciate if you find some time to investigate suppressing this. Don't block working on this if it's more difficult than it seems -- if not immediately easy to fix, then could you file an issue and downgrade the suggestion from `MachineApplicable` to `MaybeIncorrect`? :sweat_smile:

Comment (TaKO8Ki) [reply src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr:27]: > I would appreciate if you find some time to investigate suppressing this. Don't block this PR on that if it's more difficult than it seems -- if not immediately easy to fix, then could you file an issue and downgrade the suggestion from MachineApplicable to MaybeIncorrect? 😅

Ok. I will investigate it.

**Comment by TaKO8Ki** [comment]
@compiler-errors

> Anyways, this looks mostly good, except is there a reason you only suggest if we have a matching field?

There are not any specific reasons. I will implement typo suggestions for all `AssocSuggestion`.

**Comment by compiler-errors** [comment]
> There are not any specific reasons. I will implement typo suggestions for all `AssocSuggestion`.

Sounds good. If there are too many false suggestions due to enabling it for the other `AssocSuggestion`, it may justify keeping it as it is now.

**Comment by TaKO8Ki** [comment]
@compiler-errors 
I implemented typo suggestions for all `AssocSuggestion` variants.

File: src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr (hunk 1 above)
Comment (TaKO8Ki) [reply src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr:27]: I opened #97311. Can I work on this problem in a new PR?

Comment (compiler-errors) [reply src/test/ui/resolve/typo-suggestion-for-variable-with-name-similar-to-struct-field.stderr:27]: Yeah, sounds good.
